Processing GNU-style "make variables"

levitte opened the pull request:
The first step is to separate C preprocessor flags from C flags, which
has the advantage that we don't get loads of macro definitions and
inclusion directory specs when linking shared libraries, DSOs and
programs.

With that done, we can support a large set of "make variables" that we use:

```
AR              (GNU compatible)
ARFLAGS         (GNU Compatible)
AS              (GNU Compatible)
ASFLAGS         (GNU Compatible)
CC              (GNU Compatible)
CFLAGS          (GNU Compatible)
CXX             (GNU Compatible)
CXXFLAGS        (GNU Compatible)
CPP             (GNU Compatible)
CPPFLAGS        (GNU Compatible)
CPPDEFINES      List of CPP macro definitions.  Alternative for -D
CPPINCLUDES     List of CPP inclusion directories.  Alternative for -I
HASHBANGPERL    Perl invocation to be inserted after '#!' in public
                perl scripts.
LDFLAGS         (GNU Compatible)
LDLIBS          (GNU Compatible)
RANLIB          Program to generate library archive index
RC              Program to manipulate Windows resources
RCFLAGS         Flags for $(RC)
RM              (GNU Compatible)
```

Setting one of these overrides the corresponding data from our config
targets.  However, flags given directly on the configuration command
line are additional, and are therefore added to the flags coming from
one of the variables above or the config target.

Fixes #2420

levitte committed:
Processing GNU-style "make variables" - separate CPP flags from C flags

C preprocessor flags get separated from C flags, which has the
advantage that we don't get loads of macro definitions and inclusion
directory specs when linking shared libraries, DSOs and programs.

This is a step to add support for "make variables" when configuring.

levitte committed:
Processing GNU-style "make variables" - implementation

Support the following "make variables":

AR              (GNU compatible)
ARFLAGS         (GNU Compatible)
AS              (GNU Compatible)
ASFLAGS         (GNU Compatible)
CC              (GNU Compatible)
CFLAGS          (GNU Compatible)
CXX             (GNU Compatible)
CXXFLAGS        (GNU Compatible)
CPP             (GNU Compatible)
CPPFLAGS        (GNU Compatible)
CPPDEFINES      List of CPP macro definitions.  Alternative for -D
CPPINCLUDES     List of CPP inclusion directories.  Alternative for -I
HASHBANGPERL    Perl invocation to be inserted after '#!' in public
                perl scripts.
LDFLAGS         (GNU Compatible)
LDLIBS          (GNU Compatible)
RANLIB          Program to generate library archive index
RC              Program to manipulate Windows resources
RCFLAGS         Flags for $(RC)
RM              (GNU Compatible)

Setting one of these overrides the corresponding data from our config
targets.  However, flags given directly on the configuration command
line are additional, and are therefore added to the flags coming from
one of the variables above or the config target.

Fixes #2420

levitte commented:
@noloader, I'm sure you want to try this out.  And maybe @kroeckx ...

richsalz reviewed (approved):
once the CI builds pass.  And also the run-checker, maybe?

levitte commented:
Hmm, I can certainly let the run-checker do it on my laptop overnight. Good enough?

levitte commented:
Appveyor sti unhappy...l

richsalz commented:
any kind of run-checker, even just one config, is probably enough.
or you can tell me you think the travis builds are enough and i'll take that.

levitte commented:
Most of all, I need to make a VMS run as well...  but that'll be for tomorrow or monday, now is time for bed.

levitte commented:
I found that this isn't entirely done...  made it WIP for the moment being

levitte committed:
Stop having Unix defaults in Configure (partial)

Default values belong in the DEFAULT config target template, in
Configurations/00-base-templates.conf.

This isn't a complete move, but takes care of the most blatant
examples.

levitte committed:
We need Unixly defaults for config targets that don't inherit a BASE

Ideally, each config target should inherit a base to get their
platform specific defaults.  Unfortunately, that is currently not the
case, so we duplicate the Unixly defaults from the BASE_unix template
into the DEFAULT template.

levitte committed:
Make sure all our config targets inherit a BASE template

There were a small number that inherited no BASE, the now inherit
BASE_unix.

levitte commented:
Ok, this seems to work now everywhere that I can test.  Out of WIP.

richsalz reviewed (approved):
reconfirm.

levitte committed:
Add a note on Configure variable processing in NEWS and CHANGES

levitte commented:
There was a fault that Travis found (or well, clang)

Wait for the CIs to clear

levitte commented:
NOW this seems to work everywhere.  VMS cleared, Travis and Appveyor seem happy, and run-checker on my laptop hasn't complained yet (still running)...  I just had to remove two small things that didn't belong here.

levitte commented:
Now nicely rebased on a fresher master, and all fixups squashed with their intended commits

levitte commented:
Merged.  8c3bc594e0c74926bfefb84b8bae8a2fac82e465 to 3c0c6b97e1910258bfac31be407c2d8b8cded76c

levitte commented:
Thanks for the review!

bbolli commented on Configurations/10-main.conf:
```diff
@@ -192,10 +193,11 @@ my %targets = (
                                               debug   => "-O0 -g",
                                               release => "-O3 -fomit-frame-pointer"),
                                        threads("-pthread")),
+        cppflags         => add("-DL_DEBIAN"),
```
I think this should be `-DL_ENDIAN`.

levitte replied:
Heh, this is a typical case of "brain do one thing, fingers another"...  took me three hard looks before I saw the difference.  Will fix, thanks!

bbolli commented on Configurations/10-main.conf:
```diff
@@ -212,10 +214,11 @@ my %targets = (
                                               debug   => "-O0 -g",
                                               release => "-O3"),
                                        threads("-pthread")),
+        cppflags         => add("-DL_DEBIAN"),
```
I think this should be `-DL_ENDIAN`.

bbolli commented on Configurations/10-main.conf:
```diff
@@ -240,10 +243,10 @@ my %targets = (
-                                              release => "-xO5 -xdepend -xbuiltin"),
-                                       threads("-D_REENTRANT")),
+                                              release => "-xO5 -xdepend -xbuiltin")),
+        cppflags         => add("-DL_DEBIAN", threads("-D_REENTRANT")),
```
Ditto, you get the idea...

levitte commented:
@bbolli, see #5189.  Thank you!

bbolli commented:
You're welcome!
